src/persist-client/src/internal/merge.rs

```diff
@@ -48,90 +56,109 @@ impl<T> MergeTree<T> {
+            if *level_len < self.max_level_len || depth == max_level {
                 break;
             }
 
-            part = (self.merge_fn)(mem::take(level));
+            let len = mem::take(level_len);
+            self.merge_last(len);
         }
     }
 
     /// Return the contents of this merge tree, flattened into at most `max_len` parts.
-    pub fn finish(self) -> Vec<T> {
-        self.levels
-            .into_iter()
-            .reduce(|mut shallower, mut deeper| {
-                if shallower.len() + deeper.len() <= self.max_len {
-                    // Optimization: if there's enough room in the next level for everything at the
-                    // current level, add it directly.
-                    deeper.append(&mut shallower);
-                } else {
-                    // Otherwise, merge this up as if it were a full level.
-                    let merged = (self.merge_fn)(shallower);
-                    deeper.push(merged);
-                }
-                deeper
-            })
-            .expect("non-empty level array")
+    pub fn finish(mut self) -> Vec<T> {
+        let mut tail_len = 0;
+        for level_len in mem::take(&mut self.level_lens) {
+            if tail_len + level_len <= self.max_level_len {
+                // Optimization: we can combine the current level with the last level without
+                // going over our limit.
+                tail_len += level_len;
+            } else {
+                // Otherwise, perform the merge and start a new tail.
+                self.merge_last(tail_len);
+                tail_len = level_len + 1
+            }
+        }
+        assert!(self.data.len() <= self.max_level_len);
+        self.data
     }
 
     pub(crate) fn assert_invariants(&self) {
-        assert!(self.max_len >= 2, "max_len must be at least 2");
+        assert!(self.max_level_len >= 2, "max_len must be at least 2");
 
-        let (deepest, shallow) = self.levels.split_last().expect("non-empty level array");
-        for (depth, level) in shallow.iter().enumerate() {
+        assert_eq!(
+            self.data.len(),
+            self.level_lens.iter().copied().sum::<usize>(),
+            "level sizes should sum to overall len"
+        );
+        let (deepest_len, shallow) = self.level_lens.split_last().expect("non-empty level array");
+        for (depth, level_len) in shallow.iter().enumerate() {
             assert!(
-                level.len() < self.max_len,
+                *level_len < self.max_level_len,
                 "strictly less than max elements at level {depth}"
             );
         }
         assert!(
-            deepest.len() <= self.max_len,
+            *deepest_len <= self.max_level_len,
             "at most max elements at deepest level"
         );
     }
 }
 
+impl<T> Deref for MergeTree<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        &*self.data
+    }
+}
```

Comment:
neat! out of curiosity did anything motivate adding this API?

Reply:
Not something we need right now, but - one thing that's inefficient with the current Spine API is searching for the batch(es) that correspond to a particular frontier. If Spine exposed this sort of API, you could binary search your way to it, which has a nicer asymptotic behaviour...